### PR TITLE
ETH-only stake support on the client side

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs #607

Early draft. Work in progres...

### CODE REVIEW

File: pkg/chain/chain.go [hunk 1]
```diff
@@ -21,20 +21,20 @@ type Handle interface {
...
 	// to a keep.
 	RegisterAsMemberCandidate(application common.Address) error
 
-	// OnBondedECDSAKeepCreated installs a callback that is invoked when an
+	// OnKeepCreated installs a callback that is invoked when an
 	// on-chain notification of a new bonded ECDSA keep creation is seen.
-	OnBondedECDSAKeepCreated(
+	OnKeepCreated(
 		handler func(event *BondedECDSAKeepCreatedEvent),
 	) subscription.EventSubscription
```
Comment (nkuba) [inline pkg/chain/chain.go:40]: I'm not sure about these renames. The existing contract is still named `BondedECDSAKeepFactory` and event is still `BondedECDSAKeepCreatedEvent`.

File: pkg/chain/event.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 )
 
+// TODO: Rename to `KeepCreatedEvent`
```
Comment (nkuba) [inline pkg/chain/event.go:7]: I believe we need to stick to the name already defined in the `BondedECDSAKeepFactory` contract.

File: pkg/chain/chain.go [hunk 3]
```diff
@@ -65,9 +65,9 @@ type BondedECDSAKeepFactory interface {
 	GetKeepAtIndex(keepIndex *big.Int) (common.Address, error)
 }
 
-// BondedECDSAKeep is an interface that provides ability to interact with
-// BondedECDSAKeep ethereum contracts.
-type BondedECDSAKeep interface {
+// Keep is an interface that provides ability to interact with
+// keep ethereum contracts.
+type Keep interface {
```
Comment (nkuba) [inline pkg/chain/chain.go:70]: I see two possibilities:

1. We define separate interfaces for each of keep types and keep existing one to nest it in both.
```go
type FullyBackedBondedECDSAKeep interface {
  AbstractBondedECDSAKeep
}

type BondedECDSAKeep interface {
  AbstractBondedECDSAKeep
}

type ECDSAKeep interface {
// All the functions we already got in `BondedECDSAKeepFactory` interface
  OnSignatureRequested(
		keepAddress common.Address,
		handler func(event *SignatureRequestedEvent),
	) (subscription.EventSubscription, error)


	OnConflictingPublicKeySubmitted(
		keepAddress common.Address,
		handler func(event *ConflictingPublicKeySubmittedEvent),
	) (subscription.EventSubscription, error)

// etc....
}
```

2. We rename `BondedECDSAKeep` to `ECDSAKeep` (not just `Keep`).
```suggestion
type ECDSAKeep interface {
```

File: pkg/chain/ethereum/chain.go [hunk 4]
```diff
@@ -35,13 +38,12 @@ var (
...
-	bondedECDSAKeepFactoryContract *contract.BondedECDSAKeepFactory
-	blockCounter                   *blockcounter.EthereumBlockCounter
-	miningWaiter                   *ethutil.MiningWaiter
-	nonceManager                   *ethutil.NonceManager
+	config       *ethereum.Config
+	accountKey   *keystore.Key
+	client       bind.ContractBackend
+	blockCounter *blockcounter.EthereumBlockCounter
+	miningWaiter *ethutil.MiningWaiter
+	nonceManager *ethutil.NonceManager
```
Comment (nkuba) [inline pkg/chain/ethereum/chain.go:46]: What if we keep `bondedECDSAKeepFactoryContract` and add `fullyBackedBondedECDSAKeepFactoryContract`?
We will initialize them depending on the config file and if not needed leave `nil`.

```suggestion
	config       *ethereum.Config
	accountKey   *keystore.Key
	client       bind.ContractBackend
	blockCounter *blockcounter.EthereumBlockCounter
	miningWaiter *ethutil.MiningWaiter
	nonceManager *ethutil.NonceManager
	bondedECDSAKeepFactoryContract *contract.BondedECDSAKeepFactory
	fullyBackedBondedECDSAKeepFactoryContract *contract.FullyBackedECDSAKeepFactory
```

File: pkg/chain/ethereum/keep_mode.go [hunk 5]
```diff
@@ -6,30 +6,62 @@ import (
...
 	"github.com/keep-network/keep-common/pkg/subscription"
-	"github.com/keep-network/keep-core/pkg/chain"
 	eth "github.com/keep-network/keep-ecdsa/pkg/chain"
 	"github.com/keep-network/keep-ecdsa/pkg/chain/gen/contract"
 	"github.com/keep-network/keep-ecdsa/pkg/ecdsa"
 	"github.com/keep-network/keep-ecdsa/pkg/utils/byteutils"
 )
 
-var logger = log.Logger("keep-chain-eth-ethereum")
+type KeepStakeChainHandle struct {
```
Comment (nkuba) [inline pkg/chain/ethereum/keep_mode.go:22]: Not sure about this change. We would duplicate most of the `EthereumChain` for two keep types. Please see my previous comment about adding `fullyBackedBondedECDSAKeepFactoryContract` to `EthereumChain`.

**Comment by pdyraga** [comment]
Closing without merging. This feature will be addressed by staking pools.

**Pull request closed without merging**